src/hotspot/share/opto/mulnode.cpp

```diff
@@ -1296,59 +1308,160 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
+  int shift = maskShiftAmount(phase, this, bits_per_java_integer(bt));
+  if (shift == 0) {
+    return NodeSentinel;
+  }
+
+  // Check for (x & 0xFF000000) >> 24, whose mask can be made smaller.
+  // Such expressions arise normally from shift chains like (byte)(x >> 24).
+  const Node* mask = in(1);
+  if (mask->Opcode() == Op_And(bt) &&
+      (t3 = phase->type(mask->in(2))->isa_integer(bt)) &&
+      t3->is_con()) {
+    jlong maskbits = t3->get_con_as_long(bt);
+    // Convert to "(x >> shift) & (mask >> shift)"
+    Node* shr_nomask = phase->transform(RShiftNode::make(mask->in(1), in(2), bt));
+    return MulNode::make_and(shr_nomask, phase->integercon(maskbits >> shift, bt), bt);
+  }
+  return nullptr;
+}
+
+Node* RShiftNode::IdentityIL(PhaseGVN* phase, BasicType bt) {
   int count = 0;
   if (const_shift_count(phase, this, &count)) {
-    if ((count & (BitsPerJavaInteger - 1)) == 0) {
-      // Shift by a multiple of 32 does nothing
+    if ((count & (bits_per_java_integer(bt) - 1)) == 0) {
+      // Shift by a multiple of 32/64 does nothing
       return in(1);
     }
     // Check for useless sign-masking
-    if (in(1)->Opcode() == Op_LShiftI &&
+    if (in(1)->Opcode() == Op_LShift(bt) &&
         in(1)->req() == 3 &&
         in(1)->in(2) == in(2)) {
-      count &= BitsPerJavaInteger-1; // semantics of Java shifts
+      count &= bits_per_java_integer(bt)-1; // semantics of Java shifts
       // Compute masks for which this shifting doesn't change
-      int lo = (-1 << (BitsPerJavaInteger - ((uint)count)-1)); // FFFF8000
-      int hi = ~lo;               // 00007FFF
-      const TypeInt* t11 = phase->type(in(1)->in(1))->isa_int();
+      jlong lo = (-1 << (bits_per_java_integer(bt) - ((uint)count)-1)); // FFFF8000
+      jlong hi = ~lo;               // 00007FFF
+      const TypeInteger* t11 = phase->type(in(1)->in(1))->isa_integer(bt);
       if (t11 == nullptr) {
         return this;
       }
       // Does actual value fit inside of mask?
-      if (lo <= t11->_lo && t11->_hi <= hi) {
+      if (lo <= t11->lo_as_long() && t11->hi_as_long() <= hi) {
         return in(1)->in(1);      // Then shifting is a nop
       }
     }
   }
   return this;
 }
 
+const Type* RShiftNode::ValueIL(PhaseGVN* phase, BasicType bt) const {
+  const Type* t1 = phase->type(in(1));
+  const Type* t2 = phase->type(in(2));
+  // Either input is TOP ==> the result is TOP
+  if (t1 == Type::TOP) return Type::TOP;
+  if (t2 == Type::TOP) return Type::TOP;
+
+  // Left input is ZERO ==> the result is ZERO.
+  if (t1 == TypeInteger::zero(bt)) return TypeInteger::zero(bt);
+  // Shift by zero does nothing
+  if (t2 == TypeInt::ZERO) return t1;
+
+  // Either input is BOTTOM ==> the result is BOTTOM
+  if (t1 == Type::BOTTOM || t2 == Type::BOTTOM) {
+    return TypeInteger::bottom(bt);
+  }
+
+  const TypeInteger* r1 = t1->isa_integer(bt); // Handy access
+  const TypeInt* r2 = t2->isa_int(); // Handy access
```

Comment:
```suggestion
  const TypeInteger* r1 = t1->isa_integer(bt);
  const TypeInt* r2 = t2->isa_int();
```
Let's reduce the noise a little.